### DISCUSSION TITLE
关于h2协议过CDN的问题。

### DISCUSSION THREAD

**Discussion by wangdaye078** [v2fly/v2ray-core#2430] [General]
我知道很早之前有人问过，结论是当时的CDN基本都不支持h2的回源，而且贴了cloudflare的文档，明确说了不支持，但是最近我看那个文档，那一句没有了，似乎代表已经支持了h2的回源，但是我尝试h2过CDN却失败了。
我是用caddy反代的，同时上面架了个站。
使用curl -v --http2 https://test.test.top/ -o z:/test.html
看到输出是
.....
ALPN, server accepted to use h2
.....
GET / HTTP/2
.....
那代表本地确实和CDN是用的H2，caddy的日志里找到
{"remote_ip":"162.158.186.9","remote_port":"22220","proto":"HTTP/2.0","method":"GET",
经查询，上面的IP是cloudflareCDN的IP，代表CDN到caddy也是走的h2。
所有的条件都满足了，但是h2的代理却只能直连，走CDN后caddy的日志设置成debug，但什么输出都没有。不知道问题在哪。

**Comment by 1265578519** [comment]
cloudflare不支持h2回源，并且包括nginx等任何一款web软件都不支持h2回源，你在哪看到说支持的？

**Comment by wangdaye078** [comment]
1.
大家都知道GPRC过cloudflare的CDN是要去设置页专门开的，在那个设置页有个默认开的选项是
HTTP/2
Accelerates your website with HTTP/2
在下面还有一个是
HTTP/2 to Origin
Allow HTTP/2 requests between Cloudflare's edge and your origin
我想他的意思应该就是H2的回源
2.
我上面贴的输出，难道意思不是浏览器和CDN，CDN和服务器之间，用的都是H2协议吗。

**Comment by nursery01** [comment]
cloudflare 官方檔案是寫支持http2到原點，但是不支持h2的多路復用

**Comment by wangdaye078** [comment]
好吧，真是这样就没办法了。主要是ws似乎已经被盯上了，h2和grpc相比，网络上h2数据量应该比grpc要多很多，普通浏览器访问普通网站，已经有很多都是h2协议了，grpc似乎应用情况就没那么多了。所以感觉想把翻墙数据藏在普通浏览数据里的化，h2才是最好的选择。不是很懂网络，只是这么一想。

**Comment by wangdaye078** [comment]
https://developers.cloudflare.com/cache/how-to/enable-http2-to-origin
网页最后确实写了不支持多路复用，不过也写了will soon be available.查历史在去年10月的时候已经有这个了，这都过了好几个月了，还没available。开发效率有点低。希望早点支持吧。

**Comment by nursery01** [comment]
> h2和grpc相比，网络上h2数据量应该比grpc要多很多，普通浏览器访问普通网站，已经有很多都是h2协议了，grpc似乎应用情况就没那么多了。


grpc使用的就是h2，並且只要你用手機，你就有可能用到grpc，有些phone app就是grpc傳輸數據

**Reply by wangdaye078** [reply]
我知道grpc内部还是h2，只是在这个基础上又封装了一层，但是如果grpc的实际使用很少，而很多人翻墙用它，就怕被针对。你说的phone app就是grpc传输数据，这个确实是我不知道的。

**Comment by 1265578519** [comment]
ws, h2, grpc 这些都是伪装，也就等于说是压缩包，一旦解压出来，里面全是aes乱码数据，正常这些传输数据内容可不是这样的，加上vmess的tcp包有个固定头用于验证，特征比较明显自然就被检测到了，最好是使用cdn隐藏ip

**Reply by nursery01** [reply]
> ws, h2, grpc 这些都是伪装，也就等于说是压缩包，一旦解压出来，里面全是aes乱码数据，正常这些传输数据内容可不是这样的

grpc 本身就是傳亂碼數據啊，正常它傳的就是亂碼數據，除了還帶了PRI * HTTP/2.0

**Reply by wangdaye078** [reply]
ws, h2, grpc这些不是伪装，而是标准协议，只要加上TLS设置，内部传输的也是标准的TLS加密数据，如果不解密，是不知道传输的是真的网络访问数据，还是翻墙数据的，而用他们传输TLS数据也是网络标准。只要软件把其他的一个特征（比如包长，指纹等）隐藏好，墙就识别不出来（长时间访问一个IP，这种特征没办法解决）。ws+vmess是不安全的，因为会被识别，不管是vmess还是vless，还是trojan，都必须在外面套个TLS。

**Reply by nursery01** [reply]
我的意思是 grpc 不套tls的狀態下傳輸的數據就是亂碼的，因爲數據經過proto編碼，這個數據不是utf-8或者ascii，也就是說人類是讀不懂裏面的數據，grpc協定本身就是傳亂碼數據(人類看不懂)

**Reply by wangdaye078** [reply]
protobuf也就只能骗骗人眼，对于GFW来说，和明文也差不多。所以TLS还是必须的。

**Reply by nursery01** [reply]
對，可以被識別，但是假如它還沒有裝入grpc+vmess的機器學習算法，你們可以做個實驗試一試這種方式會不會被墻，特別是你小櫻，我看你說vmess+tcp被墻了，你再做一次實驗？

**Reply by 1265578519** [reply]
> 特別是你小櫻，我看你說vmess+tcp被墻了，你再做一次實驗？

刚用上vmess+tcp，几台机器全部秒墙，发生在同一个时间点，昨晚22点14分。而且都是ip地址被封，后来改ws+cdn隐藏真实ip才能继续用，但是cf延迟巨高无法用作打游戏。v2用了五年了，也提交给v2修复过pr，我不至于犯蠢。
![image](https://user-images.githubusercontent.com/6442439/226870325-d4824604-fedc-4081-bac6-ce26f17a3d65.png)

![image](https://user-images.githubusercontent.com/6442439/226870768-132f1b72-acd6-4b3a-99b2-6a690a9648cf.png)

**Reply by nursery01** [reply]
> 刚用上vmess+tcp，几台机器全部秒墙，发生在同一个时间点

這。。。你有沒有用臟ip？我是覺得很奇怪爲什麽有些人會被墻，明明都用差不多的技術，這個gfw真的是個迷啊

**Reply by 1265578519** [reply]
> 這。。。你有沒有用臟ip？

我不懂繁体，不明白你在说什么，我使用了翻译软件后还是没看明白
![image](https://user-images.githubusercontent.com/6442439/226875054-48718956-ba67-4476-a65f-b840003f0856.png)

**Reply by wangdaye078** [reply]
臟ip，大概意思是以前被人用来做过翻墙，然后被标记了的。或者是某些常被用来翻墙的热门的VPS商

**Reply by nursery01** [reply]
樓上正確

**Reply by 1265578519** [reply]
> 臟ip，大概意思是以前被人用来做过翻墙，然后被标记了的。或者是某些常被用来翻墙的热门的VPS商

不至于，这些服务器都是自己用的三年的ip了，早期一直是ws稳定跑几年，从上个月底开始就不行了一直封ws端口，需要换端口才能用，tls tcp等都试过了是封ip，看来vmess已经被通杀了
![image](https://user-images.githubusercontent.com/6442439/226876290-e70e8331-1873-45c4-b528-38f361d2899e.png)

**Reply by wangdaye078** [reply]
至于用一样的技术，某些人会被墙，某些人不会，我有一个想法。20年前有个游戏叫奇迹，漏洞层出不穷，这个版本封堵了，下个版本又出来了，最后发现是他们一套服务器就一个程序，要封堵漏洞就要增加服务器判断，增加了CPU压力，下个版本一增加新功能，压力继续增加，服务器扛不住，就只能减少一些代码复杂度，所以老漏洞就又出来了。GFW可能也一样，如果所有数据都过滤，那可能算力不够，所以对某些IP有重点照顾，比如搬瓦工这种热门VPS商的IP段（我猜的），或者某人被抓后就重点照顾相近网段的IP。反正就是用尽量少的算力，达到一个让上头满意的结果就好了。另外有时候会明显感觉用某种方法的被抓了一批，然后过段时间又可以用了，大概也可能是为了平衡算力做的妥协。所以我们能做的就是把自己的数据混入过墙的流量最大的那种数据里面，而且尽量不出现很容易被抓出来的特征，那应该基本就不会被抓了。

**Reply by nursery01** [reply]
> 不至于，这些服务器都是自己用的三年的ip了，早期一直是ws稳定跑几年，从上个月底开始就不行了一直封ws端口，需要换端口才能用，tls tcp等都试过了是封ip，看来vmess已经被通杀了 !

原來如此，你用ws當然被墻啦，近一年跑到github上說被墻的大部分都是用ws，不用人工智慧，用固定的代碼就能識別ws
你被識別完你的ip就已經是惡名ip了，你再換技術也是會被秒封

**Reply by 1265578519** [reply]
> 原來如此，你用ws當然被墻啦，近一年跑到github上說被墻的大部分都是用ws，不用人工智慧，用固定的代碼就能識別ws 你被識別完你的ip就已經是惡名ip了，你再換技術也是會被秒封

换个新服务器新的ip，早上装了后还没用，纯tcp的，晚上回家电脑刚开机，一分钟后ip就嘎了。服务器日志就断了。日志上没有不明来源访问，很明显是给精准检测识别了
![image](https://user-images.githubusercontent.com/6442439/227263856-2732eca6-21aa-40f1-90f0-5f101c380b48.png)

**Reply by wangdaye078** [reply]
过CDN的话，试试GRPC+TLS，不过CDN的话，试试H2+CDN，另外v2rayN，passwall等好像还不支持V2RAY的UTLS设置，可以用XRAY试试。另外如果用来翻墙打游戏的话，可以试试香港的腾讯云，反正我的那机器一直没被封过，只是别拿它去发什么时政相关的话。

**Reply by wangdaye078** [reply]
如果不用反代的话，还可以试试XRAY的新协议或者QUIC+TLS，纯TCP不加TLS的话，就是裸奔，加TLS的话，反正我想不出有多少应用是用TCP协议跑TLS数据。怎么也得模仿个HTTP框架下的东西吧。

**Reply by 1265578519** [reply]
> 反正我想不出有多少应用是用TCP协议跑TLS数据。

bt下载使用openssl跑TCP的TLS，但是v2不具备这个功能，只能用kcp的utp，没有tcp的bt伪装，，，用的就是腾讯云，cod玩着玩着就嘎了，每台机器也只能换一次ip地址，tcp和https是不能用了，百分百精准识别秒嘎掉封ip
![image](https://user-images.githubusercontent.com/6442439/227475679-edc145f4-fba7-47ca-8b5a-1e185da31652.png)
今天换回了ws还好，这样只会封端口，而且ws不容易被检测到，改个端口就继续用了。

**Reply by wangdaye078** [reply]
如果V2能模拟BT协议倒是可以期待下，共同点很多，长时间连接，数据断断续续，有时候流量大有时候流量小，上传少，下载多。我的腾讯云用了快3年了，一直没遇到问题。

**Reply by android1122** [reply]
> 过CDN的话，试试GRPC+TLS，不过CDN的话，试试H2+CDN，另外v2rayN，passwall等好像还不支持V2RAY的UTLS设置，可以用XRAY试试。另外如果用来翻墙打游戏的话，可以试试香港的腾讯云，反正我的那机器一直没被封过，只是别拿它去发什么时政相关的话。

这个没明白，h2+cdn不是还没支持，现在只能是ws+cdn

**Reply by wangdaye078** [reply]
写错了，是不过CDN的话，试试H2+TLS

**Reply by android1122** [reply]
> 写错了，是不过CDN的话，试试H2+TLS

在我的认知里，CDN最大的作用是用来救IP的，所以一直都是ws+tls+cdn，用到现在也没啥问题，就是速度一言难尽

**Comment by 1265578519** [comment]
之前上面说的脏ip说法，应该不攻自破了，测试在ws墙端口的机器上，连续换了几次端口，端口都被墙掉，故意这样把ip弄脏，然后转换其它协议继续测试。
目前转了双双几天，一点事情都没有了，所以脏ip这个说法不成立。

**Comment by 1265578519** [comment]
https://web.archive.org/web/20230331092744/https://www.aodun.com.cn/security_info/vpn
应该是这个原因了，怪不得我这秒墙，估计当地运营商防火墙已经部署上傲盾了，精准识别v2

**Comment by rhjdvsgsgks** [comment]
ngnix 好像支持 h2c 回源了
https://trac.nginx.org/nginx/ticket/816
